### ISSUE TITLE
Package request: code-cursor

### ISSUE THREAD

**Issue by Anifyuli** [NixOS/nixpkgs#309541]
**Project description**
This is Code OSS fork with builtin AI code generator

**Metadata**

* homepage URL: https://cursor.sh/
* source URL: https://github.com/getcursor/cursor (I haven't idea how to get direct link from it)
* license: mit, unfree (I guess)
* platforms: linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by gebeer** [comment]
They only distribute cursor editor as Appimage. It would really be great to have this packaged for Nix.
The Appimage is updated very frequently. So the nix package name should include the version number, sth like `cursor-0.33`

**Comment by haras-unicorn** [comment]
```nix
pkgs.appimageTools.wrapType2 {
          pname = "cursor";
          version = "0.1.0";

          src = pkgs.fetchurl {
            url = "https://downloader.cursor.sh/linux/appImage/x64";
            hash = "sha256-Fsy9OVP4vryLHNtcPJf0vQvCuu4NEPDTN2rgXO3Znwo=";
          };
        }
```

heres a start if u wanna try it
if it complains change the hash to the thing it expects
idk if there are versioned urls for this thing so that the hash is not a problem

on another note this thing is so slow
like no extensions (except the builtin vim mode) and it is slow and stutters on my machine and idk why
vscode works fast compared to it
but idk mby im doing something wrong

**Comment by gebeer** [comment]
> ```nix
> pkgs.appimageTools.wrapType2 {
>           pname = "cursor";
>           version = "0.1.0";
> 
>           src = pkgs.fetchurl {
>             url = "https://downloader.cursor.sh/linux/appImage/x64";
>             hash = "sha256-Fsy9OVP4vryLHNtcPJf0vQvCuu4NEPDTN2rgXO3Znwo=";
>           };
>         }
> ```
> 
> heres a start if u wanna try it if it complains change the hash to the thing it expects idk if there are versioned urls for this thing so that the hash is not a problem
> 
> on another note this thing is so slow like no extensions (except the builtin vim mode) and it is slow and stutters on my machine and idk why vscode works fast compared to it but idk mby im doing something wrong

Thank you for the suggestion.

I've gone another route to avoid the hash thing because the AppImage is being updated quite frequently.
I have a module cursor.nix with a shell script:
```nix
{pkgs, homeDir ? pkgs.lib.getHome} :
pkgs.writeShellScriptBin "cursor" ''
    # Find the latest cursor AppImage in ~/Applications
    echo "Home Directory: ${homeDir}"
    cursor_app="$(find ${homeDir}/Applications -maxdepth 1 -name 'cursor*.AppImage' | sort | tail -n 1)"
    if [[ -f "$cursor_app" ]]; then
        # Execute the AppImage if found
        appimage-run "$cursor_app" "$@"
    else
        echo "Cursor AppImage not found or not executable in ~/Applications."
        exit 1
    fi
''
```
and in my packages.nix I install it:
```nix
    (callPackage ./shellscripts/cursor.nix {
      # inherit pkgs;
      homeDir = "${homeDir}";
    })
```

When I get prompted for an update in the cursor app, I download the latest AppImage and the shell script takes care of the rest.
Still a manual download step involved. But imo better than having to rebuild twice because of the hash error.

**Comment by haras-unicorn** [comment]
> > ```nix
> > pkgs.appimageTools.wrapType2 {
> >           pname = "cursor";
> >           version = "0.1.0";
> > 
> >           src = pkgs.fetchurl {
> >             url = "https://downloader.cursor.sh/linux/appImage/x64";
> >             hash = "sha256-Fsy9OVP4vryLHNtcPJf0vQvCuu4NEPDTN2rgXO3Znwo=";
> >           };
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > heres a start if u wanna try it if it complains change the hash to the thing it expects idk if there are versioned urls for this thing so that the hash is not a problem
> > on another note this thing is so slow like no extensions (except the builtin vim mode) and it is slow and stutters on my machine and idk why vscode works fast compared to it but idk mby im doing something wrong
> 
> Thank you for the suggestion.
> 
> I've gone another route to avoid the hash thing because the AppImage is being updated quite frequently. I have a module cursor.nix with a shell script:
> 
> ```nix
> {pkgs, homeDir ? pkgs.lib.getHome} :
> pkgs.writeShellScriptBin "cursor" ''
>     # Find the latest cursor AppImage in ~/Applications
>     echo "Home Directory: ${homeDir}"
>     cursor_app="$(find ${homeDir}/Applications -maxdepth 1 -name 'cursor*.AppImage' | sort | tail -n 1)"
>     if [[ -f "$cursor_app" ]]; then
>         # Execute the AppImage if found
>         appimage-run "$cursor_app" "$@"
>     else
>         echo "Cursor AppImage not found or not executable in ~/Applications."
>         exit 1
>     fi
> ''
> ```
> 
> and in my packages.nix I install it:
> 
> ```nix
>     (callPackage ./shellscripts/cursor.nix {
>       # inherit pkgs;
>       homeDir = "${homeDir}";
>     })
> ```
> 
> When I get prompted for an update in the cursor app, I download the latest AppImage and the shell script takes care of the rest. Still a manual download step involved. But imo better than having to rebuild twice because of the hash error.

yea makes sense - im not using it actively and just wanted to quickly check it out

**Comment by AurelienConte** [comment]
Any update on this ?

Using : https://gist.github.com/lukalot/fcbf3216ad13b8303ab0947af0d5abd5 actually, i think this is a good start but i'm looking a way to deal with downloading the latest and breaking the sha256 all times at every cursor updates

**Comment by sarahec** [comment]
Here's a way to write the derivation:

https://github.com/NixOS/nixpkgs/blob/8da188f8e754473cb2f5a922bef75a734bfa0fd6/pkgs/applications/networking/instant-messengers/beeper/default.nix#L15

**Comment by sarahec** [comment]
I added a PR. You may want to upvote or review it.

**Comment by sarahec** [comment]
P.S. I'm also looking for co-maintainers

**Comment by sarahec** [comment]
Update: I've had one reviewer go over the code and I'm awaiting a second who can also commit it. 

Once this is done, I can issue a backport to stable.

**Comment by sarahec** [comment]
Sorry friends, we're just waiting on it to be merged.

**Comment by sarahec** [comment]
Quick note here: please don't add additional comments to the PR unless you are a reviewer. Long conversation trails make PRs are much harder to review and significantly reduce the odds they'll get merged.

I've tried moving the request up to get higher visibility, and I suspect the large number of comments are scaring off the committers.

**Comment by sarahec** [comment]
Finally, it's merged! Enjoy, and please remember that application bugs should be reported on the cursor forum.

**Comment by kylecarbs** [comment]
If anyone has problems with the merged flake, you can try [ours out](https://github.com/coder/cursor-arm). It also works for arm!

**Comment by sarahec** [comment]
FYI: We are up on the unstable channel. Enjoy!

**Comment by Oudwins** [comment]
@sarahec  I am unable to update it. Is the package broken or is it an issue on my end? Anyone else have this issue?

This is what the output of switching shows:
```nix
❯ sudo nixos-rebuild switch --flake .#
building the system configuration...
error: builder for '/nix/store/9dpwjd8lv1h60ikamd4hcipc3pb5ah8j-cursor-0.42.4-build-2410291z3bdg1dy-x86_64.AppImage.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://download.todesktop.com/230313mzl4w4u92/cursor-0.42.4-build-2410291z3bdg1dy-x86_64.AppImage
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >  31  155M   31 48.7M    0     0   185k      0  0:14:18  0:04:29  0:09:49 60015
       > curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
       > error: cannot download cursor-0.42.4-build-2410291z3bdg1dy-x86_64.AppImage from any mirror
       For full logs, run 'nix log /nix/store/9dpwjd8lv1h60ikamd4hcipc3pb5ah8j-cursor-0.42.4-build-2410291z3bdg1dy-x86_64.AppImage.drv'.
error: 1 dependencies of derivation '/nix/store/zjx38r9imhm49xg60idv1jz515k9vkii-cursor-0.42.4-extracted.drv' failed to build
error: 1 dependencies of derivation '/nix/store/6fln155vmhccwpn07w4djppqg6jbs0ws-cursor-0.42.4.drv' failed to build
error: 1 dependencies of derivation '/nix/store/a233i9dpalyyv40v5dq32qcv2na6gzwp-home-manager-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/iwi8f5aim6wc2fpzy9m0l5xkdahai5cl-home-manager-generation.drv' failed to build
error: 1 dependencies of derivation '/nix/store/4da7b3svsjk018khg1fna5xkb7pr78zc-user-environment.drv' failed to build
error: 1 dependencies of derivation '/nix/store/3s2869j6chfaqng63hjfk27rz70kqkbc-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/gr8qdc61dccnnm8lylfhzm20cf8f20g3-nixos-system-nixos-24.05.20241110.9256f7c.drv' failed to build
```

**Comment by sarahec** [comment]
It looks like they released a new version *today* and deleted the old one.
I'm filing a new PR.

**Comment by sarahec** [comment]
I need a reviewer and committer for #356008

**Comment by sarahec** [comment]
> @sarahec I am unable to update it. Is the package broken or is it an issue on my end? Anyone else have this issue?

If you're using flakes, I'd revert the flake.lock change. If you're on Matrix, nixpkgs-bot can watch for the PR to be merged.
